[CSS ZOOM] Apply zoom factor to border spacing

nullhook opened the pull request:
#### ce49a05a919733fe9f56f50e17bb60bb56c984d8
<pre>
[CSS ZOOM] Apply zoom factor to border spacing
<a href="https://bugs.webkit.org/show_bug.cgi?id=300026">https://bugs.webkit.org/show_bug.cgi?id=300026</a>
<a href="https://rdar.apple.com/161816440">rdar://161816440</a>

Reviewed by Antti Koivisto.

We apply the zoom factor at evaluation-time for the border-spacing property.

The R bit in Length&lt;R, V&gt; is now annotated as Length&lt;CSS::NonnegativeUnzoomed&gt;,
so that zooming is deferred until evaluation-time rather than being applied during style resolution.

Here, evaluation-time means: the phase after style resolution, during rendering/layout.

Tests: imported/w3c/web-platform-tests/css/css-viewport/zoom/border-spacing-without-border-ref.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/border-spacing-without-border.html
       imported/w3c/web-platform-tests/svg/styling/border-spacing-svg-zoom-computed.html
       imported/w3c/web-platform-tests/svg/styling/border-spacing-svg-zoom-table-display-ref.html
       imported/w3c/web-platform-tests/svg/styling/border-spacing-svg-zoom-table-display.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-spacing-without-border-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-spacing-without-border-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-spacing-without-border.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/border-spacing-svg-zoom-computed-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/border-spacing-svg-zoom-computed.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/border-spacing-svg-zoom-table-display-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/border-spacing-svg-zoom-table-display-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/border-spacing-svg-zoom-table-display.html: Added.
* Source/WebCore/accessibility/AXTableHelpers.cpp:
(WebCore::AXTableHelpers::isDataTableWithTraversal):
* Source/WebCore/layout/formattingContexts/table/TableFormattingState.cpp:
(WebCore::Layout::ensureTableGrid):
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::styleDidChange):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/style/values/non-standard/StyleWebKitBorderSpacing.h:

Canonical link: <a href="https://commits.webkit.org/301214@main">https://commits.webkit.org/301214@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4ce65429c1a6658e1b47e9ca0ad6ca1a073e8dfe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124745 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44417 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35152 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131591 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76664 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/eba11c69-c1cc-42b9-ac1b-8e9d96db39eb) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126622 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45121 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52987 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94911 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/63187 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e323dc26-4bdc-4230-aaa9-386de74d8479) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127699 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35989 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111563 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75480 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/73af0e96-222a-451b-8b23-ece8a8350425) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34920 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29739 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75067 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105746 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29948 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134256 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51593 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39402 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103388 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52006 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107783 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103159 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/26380 "The change is no longer eligible for processing.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48527 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26805 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48611 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19617 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51466 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57264 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50860 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54215 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52554 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8e707d63406707b9c2c64b3ec134b29a179785f7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51692

weinig commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -2694,7 +2694,7 @@ EnumeratingVisibleNetworkInterfacesEnabled:
   category: css
   type: bool
-  status: unstable
+  status: testable
```
Why are you flipping this?

weinig reviewed (changes requested):
This still needs testing of the SVG rules cases.

nullhook commented:
> This still needs testing of the SVG rules cases

border-spacing only applies to table elements with border-collapse: separate. this is actually a nuanced edge case, its technically possible but i'm not sure if it's worth testing?

i haven't fully formalized my thought but an idea was house a table under foreignObject within SVG tree.

```
<svg style="zoom: 2.0">
    <foreignObject width="200" height="100">
      <table style="border-spacing: 10px; border-collapse: separate;">
        <tr>
          <td>Cell 1</td>
          <td>Cell 2</td>
        </tr>
      </table>
    </foreignObject>
</svg>
```

nullhook commented:
```
<svg style="border-spacing: 10px 20px;">
  <rect id="target" width="200" height="100" x="10" y="10" rx="20" ry="20" fill="blue" />
</svg>
```

this makes no sense to me. border-spacing only has meaning when there are adjacent cells to separate under a table structure. i have no idea what above is trying to achieve.

nullhook committed:
[CSS ZOOM] Apply zoom factor to border spacing
https://bugs.webkit.org/show_bug.cgi?id=300026
rdar://161816440

Reviewed by Antti Koivisto.

We apply the zoom factor at evaluation-time for the border-spacing property.

The R bit in Length<R, V> is now annotated as Length<CSS::NonnegativeUnzoomed>,
so that zooming is deferred until evaluation-time rather than being applied during style resolution.

Here, evaluation-time means: the phase after style resolution, during rendering/layout.

Tests: imported/w3c/web-platform-tests/css/css-viewport/zoom/border-spacing-without-border-ref.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/border-spacing-without-border.html
       imported/w3c/web-platform-tests/svg/styling/border-spacing-svg-zoom-computed.html
       imported/w3c/web-platform-tests/svg/styling/border-spacing-svg-zoom-table-display-ref.html
       imported/w3c/web-platform-tests/svg/styling/border-spacing-svg-zoom-table-display.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-spacing-without-border-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-spacing-without-border-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-spacing-without-border.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/border-spacing-svg-zoom-computed-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/border-spacing-svg-zoom-computed.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/border-spacing-svg-zoom-table-display-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/border-spacing-svg-zoom-table-display-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/styling/border-spacing-svg-zoom-table-display.html: Added.
* Source/WebCore/accessibility/AXTableHelpers.cpp:
(WebCore::AXTableHelpers::isDataTableWithTraversal):
* Source/WebCore/layout/formattingContexts/table/TableFormattingState.cpp:
(WebCore::Layout::ensureTableGrid):
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::styleDidChange):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/style/values/non-standard/StyleWebKitBorderSpacing.h:

Canonical link: https://commits.webkit.org/301214@main
